### PR TITLE
Convert to V2 addon

### PR DESCRIPTION
**Pull request by herzzanu**

### CODE REVIEW

**Comment by SkoebaSteve** [comment]
@herzzanu ok to wait with merging this until we release v2?

**Comment by herzzanu** [comment]
@SkoebaSteve are we good to merge at this point or waiting for something else? Happy to help and unblock the situation.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -46,7 +50,7 @@ jobs:
       - name: Install Dependencies
         run: yarn install --frozen-lockfile
       - name: Run Tests
-        run: yarn test:ember
+        run: yarn test
```
Comment (anas7asia) [inline .github/workflows/ci.yml:53]:
```suggestion
        run: yarn test:ember
        working-directory: ./test-app
```

Like this we won't need to run the lint job twice

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -77,4 +80,5 @@ jobs:
       - name: Install Dependencies
         run: yarn install --frozen-lockfile
       - name: Run Tests
-        run: ./node_modules/.bin/ember try:one ${{ matrix.try-scenario }}
+        run: yarn ember try:one ${{ matrix.try-scenario }}
+        working-directory: ./test-app
```
Comment (anas7asia) [inline .github/workflows/ci.yml:84]:
```suggestion
        working-directory: test-app
```
😄

**Comment by anas7asia** [comment]
Maybe we can add a prettier configuration for the addon as well?

File: ember-launch-darkly/LICENSE.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
+The MIT License (MIT)
```
Comment (anas7asia) [inline ember-launch-darkly/LICENSE.md:1]: This file can be left on the root level, it is copied by rollup when publishing

File: ember-launch-darkly/package.json [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    }
+  },
+  "exports": {
+    ".": "./dist/index.js",
+    "./*": {
+      "default": "./dist/*.js"
+    },
+    "./test-support": "./dist/test-support/index.js",
+    "./-sdk": "./dist/-sdk/index.js",
+    "./-sdk/context": "./dist/-sdk/context.js",
```
Comment (anas7asia) [inline ember-launch-darkly/package.json:73]: These are private files and they won't be imported by addon users, right?

File: ember-launch-darkly/README.md
Comment (anas7asia) [inline ember-launch-darkly/README.md:1]: README can be left on the root level as well, it is copied by rollup when publishing:
```
    // Copy Readme and License into published package
    copy({
      targets: [
        { src: '../README.md', dest: '.' },
        { src: '../LICENSE.md', dest: '.' },
      ],
    }),
```

File: ember-launch-darkly/package.json [hunk 5]
```diff
@@ -0,0 +1,79 @@
+{
+  "name": "ember-launch-darkly",
+  "version": "2.2.0",
```
Comment (anas7asia) [inline ember-launch-darkly/package.json:3]: Isn't it v3.0.0 now?

File: ember-launch-darkly/package.json [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    "lint:js": "eslint . --cache",
+    "lint:hbs:fix": "ember-template-lint . --fix --no-error-on-unmatched-pattern",
+    "lint:js:fix": "eslint . --fix",
+    "start": "rollup --config --watch",
+    "test": "echo 'A v2 addon does not have tests, run tests in test-app'",
+    "prepack": "rollup --config"
+  },
+  "dependencies": {
+    "@embroider/addon-shim": "^1.0.0",
+    "launchdarkly-js-client-sdk": "^2.24.2",
```
Comment (anas7asia) [inline ember-launch-darkly/package.json:33]: It was upgraded to ^3.1.1

File: ember-launch-darkly/package.json [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+  },
+  "devDependencies": {
+    "@babel/core": "^7.21.3",
+    "@babel/plugin-proposal-class-properties": "^7.18.6",
+    "@babel/plugin-proposal-decorators": "^7.21.0",
+    "@embroider/addon-dev": "^3.0.0",
+    "@rollup/plugin-babel": "^6.0.3",
+    "babel-eslint": "^10.1.0",
+    "concurrently": "^7.6.0",
+    "ember-source": "4.11.0",
```
Comment (anas7asia) [inline ember-launch-darkly/package.json:44]: Do you think we can add it as a peer dep? Example: https://github.com/adopted-ember-addons/ember-launch-darkly/blob/master/package.json#L77

**Comment by herzzanu** [comment]
> Maybe we can add a prettier configuration for the addon as well?

That's a great idea. I prefer to add it in a separate PR, once we merge, since it was not there before.

File: ember-launch-darkly/rollup.config.mjs [hunk 8]
```diff
@@ -0,0 +1,41 @@
+import copy from 'rollup-plugin-copy';
+import { Addon } from '@embroider/addon-dev/rollup';
+
```
Comment (anas7asia) [inline ember-launch-darkly/rollup.config.mjs:3]: Do you think we should add babel config?

File: ember-launch-darkly/rollup.config.mjs [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+export default {
+  // This provides defaults that work well alongside `publicEntrypoints` below.
+  // You can augment this if you need to.
+  output: addon.output(),
+
+  plugins: [
+    // These are the modules that users should be able to import from your
+    // addon. Anything not listed here may get optimized away.
+    addon.publicEntrypoints(['**/*.js']),
```
Comment (anas7asia) [inline ember-launch-darkly/rollup.config.mjs:17]: Do we need index.js as well?

```suggestion
    addon.publicEntrypoints(['index.js', '**/*.js']),
```

File: ember-launch-darkly/rollup.config.mjs [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+    // These are the modules that should get reexported into the traditional
+    // "app" tree. Things in here should also be in publicEntrypoints above, but
+    // not everything in publicEntrypoints necessarily needs to go here.
+    addon.appReexports(['helpers/**/*.js']),
+
```
Comment (anas7asia) [inline ember-launch-darkly/rollup.config.mjs:23]: Don't we need dependencies config?

```suggestion
    // Follow the V2 Addon rules about dependencies. Your code can import from
    // `dependencies` and `peerDependencies` as well as standard Ember-provided
    // package names.
    addon.dependencies(),
```

File: package.json [hunk 11]
```diff
@@ -1,89 +1,28 @@
 {
-  "name": "ember-launch-darkly",
-  "version": "2.2.0",
-  "description": "A modern Ember addon to wrap the Launch Darkly service",
-  "keywords": [
-    "ember-addon"
+  "private": true,
```
Comment (anas7asia) [inline package.json:2]: I think we can keep the name and the version in this package.json as well

File: test-app/.editorconfig [hunk 12]
```diff
@@ -0,0 +1,19 @@
+# EditorConfig helps developers define and maintain consistent
```
Comment (anas7asia) [inline test-app/.editorconfig:1]: Should it stay at the root level? 🤔

File: test-app/.prettierrc.js [hunk 13]
```diff
@@ -0,0 +1,12 @@
+'use strict';
+
+module.exports = {
```
Comment (anas7asia) [inline test-app/.prettierrc.js:3]: Should it have settings for endOfFile and singleQuote like it did: https://github.com/adopted-ember-addons/ember-launch-darkly/blob/master/.prettierrc?

File: test-app/README.md [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+* [Git](https://git-scm.com/)
+* [Node.js](https://nodejs.org/) (with npm)
+* [Ember CLI](https://cli.emberjs.com/release/)
+* [Google Chrome](https://google.com/chrome/)
+
+## Installation
+
+* `git clone <repository-url>` this repository
+* `cd test-app`
+* `npm install`
```
Comment (anas7asia) [inline test-app/README.md:19]:
```suggestion
* `yarn install`
```

File: test-app/README.md [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+
+### Running Tests
+
+* `ember test`
+* `ember test --server`
+
+### Linting
+
+* `npm run lint`
+* `npm run lint:fix`
```
Comment (anas7asia) [inline test-app/README.md:39]:
```suggestion
* `yarn lint`
* `yarn lint:fix`
```

File: test-app/app/identified-user/controller.js [hunk 16]
```diff
@@ -1,10 +1,9 @@
 import Controller from '@ember/controller';
-import { capitalize } from '@ember/string';
 
 import { variation } from 'ember-launch-darkly';
 
 export default class IdentifiedUserController extends Controller {
   get shape() {
-    return capitalize(variation('shape'));
+    return variation('shape');
```
Comment (anas7asia) [inline test-app/app/identified-user/controller.js:7]: Is `capitalized` not needed anymore?

File: test-app/app/index/template.hbs [hunk 17]
```diff
@@ -0,0 +1,10 @@
+<br />
+<br />
+
+{{#let (component (ensure-safe-component this.shape)) as |Shape|}}
```
Comment (anas7asia) [inline test-app/app/index/template.hbs:4]: Has the template changed?

File: test-app/config/ember-cli-update.json [hunk 18]
```diff
@@ -6,13 +6,13 @@
...
-          "codemodsSource": "ember-addon-codemods-manifest@1",
+          "name": "app",
+          "outputRepo": "https://github.com/ember-cli/ember-new-output",
+          "codemodsSource": "ember-app-codemods-manifest@1",
           "isBaseBlueprint": true,
           "options": [
-            "--yarn",
-            "--no-welcome"
+            "--no-welcome",
+            "--ci-provider=github"
```
Comment (anas7asia) [inline test-app/config/ember-cli-update.json:15]:
```suggestion
            "--yarn"
```

File: test-app/config/ember-try.js [hunk 19]
```diff
@@ -23,18 +23,8 @@ module.exports = async function () {
           },
         },
       },
-      {
-        name: 'ember-lts-4.4',
```
Comment (anas7asia) [inline test-app/config/ember-try.js:27]: Should we keep 4.4 scenario? I seems we still support it

File: test-app/config/ember-try.js [hunk 20]
```diff
@@ -52,33 +41,26 @@ module.exports = async function () {
...
         npm: {
           devDependencies: {
-            'ember-source': '~3.28.0',
+            'ember-source': await getChannelURL('release'),
           },
-          ember: {
-            edition: 'classic',
+        },
+      }),
+      embroiderOptimized({
```
Comment (anas7asia) [inline test-app/config/ember-try.js:57]: Can't we leave Embroider scenario's config empty?

File: test-app/package.json [hunk 21]
```diff
@@ -0,0 +1,71 @@
+{
+  "name": "test-app",
+  "version": "0.0.0",
+  "private": true,
+  "description": "Small description for test-app goes here",
+  "repository": "",
+  "license": "MIT",
+  "author": "",
```
Comment (anas7asia) [inline test-app/package.json:8]:
```suggestion
  "author": "",
```
As it's private, we can remove it completely, right?

File: test-app/testem.js [hunk 22]
```diff
@@ -11,9 +11,7 @@ module.exports = {
       ci: [
         // --no-sandbox is needed when running Chrome inside a container
         process.env.CI ? '--no-sandbox' : null,
-        process.platform === 'darwin' ? '--crash-dumps-dir=/tmp' : null, // Fix for https://superuser.com/questions/1292863/chrome-crashpad-crashes-on-xattr
```
Comment (anas7asia) [inline test-app/testem.js:14]: Should we remove it in a separate PR?

File: ember-launch-darkly/package.json [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+      "default": "./dist/*.js"
+    },
+    "./test-support": "./dist/test-support/index.js",
+    "./-sdk": "./dist/-sdk/index.js",
+    "./-sdk/context": "./dist/-sdk/context.js",
+    "./-sdk/identify": "./dist/-sdk/identify.js",
+    "./-sdk/initialize": "./dist/-sdk/initialize.js",
+    "./-sdk/variation": "./dist/-sdk/variation.js",
+    "./addon-main.js": "./addon-main.js"
+  }
```
Comment (anas7asia) [inline ember-launch-darkly/package.json:77]:
```suggestion
  },
  "peerDependencies": {
    "ember-source": "^3.28.0 || ^4.0.0"
  },
```

We need to keep the peerDependencies that was in v1, right?

Comment (anas7asia) [reply ember-launch-darkly/package.json:80]: Alright, thanks!

**Comment by tniezurawski** [comment]
@herzzanu This might be a bit odd question but I wanted to help the community to move some addons from V1 to V2 so... how did you know what to do? 😅 Is there some "v1 to v2 TODO list" resource?

**Comment by herzzanu** [comment]
> @herzzanu This might be a bit odd question, but I wanted to help the community to move some addons from V1 to V2 so... how did you know what to do? 😅 Is there some "v1 to v2 TODO list" resource?

@tniezurawski, that's great to hear 🙌 Here's a non-exhaustive list of resources that could help:

#### Guides

- https://github.com/embroider-build/embroider/blob/main/docs/porting-addons-to-v2.md
- https://www.kaliber5.de/en/blog/v2-addon_en
- https://medium.com/@phndiaye/migrating-an-ember-addon-to-the-v2-format-1d62e45384d3

#### Tools

- https://github.com/NullVoxPopuli/ember-addon-migrator
- https://github.com/ijlee2/ember-codemod-v1-to-v2

#### A couple of v2 addons that could serve as inspiration

- https://github.com/NullVoxPopuli/ember-resources
- https://github.com/CrowdStrike/ember-headless-form
- https://github.com/simonihmig/ember-stargate
- https://github.com/ember-cli/ember-welcome-page
- https://github.com/ijlee2/ember-container-query
- https://github.com/qonto/ember-amount-input
- https://github.com/qonto/ember-autofocus-modifier

The list of V2 add-ons can go on. It's also a good idea to find the specific commits that helped migrate these addons to understand the nature of the changes. Hope it helps. Happy to help if you need anything specific on any of the projects you plan to work on.

**Pull request merged**